### PR TITLE
DecompressedSize added to Resource table

### PR DESCRIPTION
**Pull request by rajithaalurims**
## Description
A new nullable column DecompressedSize added to Resource table.
A new table type is created to support backward and forward compatibility.
Stored procedures are changed to use data from either old or new data type depending on schema version supported on the instance.

## Related issues
Addresses [issue #[156583](https://microsofthealth.visualstudio.com/Health/_workitems/edit/156583)].

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- When changing or adding behavior, if your code modifies the system design or changes design assumptions, please create and include an [ADR](https://github.com/microsoft/fhir-server/blob/main/docs/arch).
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/main/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/main/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql [hunk 1]
```diff
@@ -159,10 +196,20 @@ BEGIN TRY
       --EXECUTE dbo.LogEvent @Process=@SP,@Mode=@Mode,@Status='Info',@Start=@st,@Rows=@AffectedRows,@Text='Old rows'
     END
 
-    INSERT INTO dbo.Resource 
-           ( ResourceTypeId, ResourceId, Version, IsHistory, ResourceSurrogateId, IsDeleted, RequestMethod, RawResource, IsRawResourceMetaSet, SearchParamHash,  TransactionId )
-      SELECT ResourceTypeId, ResourceId, Version, IsHistory, ResourceSurrogateId, IsDeleted, RequestMethod, RawResource, IsRawResourceMetaSet, SearchParamHash, @TransactionId
-        FROM @Resources
+    IF @HasDecompressedSize = 1
+    BEGIN
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql]: No need for BEGIN/END

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql [hunk 2]
```diff
@@ -159,10 +196,20 @@ BEGIN TRY
...
+        INSERT INTO dbo.Resource
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql]: Indentation does not match the rest of code

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+  EXECUTE dbo.LogEvent @Process=@SP,@Mode=@Mode,@Status='Error',@Start=@st;
+
+  IF @RaiseExceptionOnConflict = 1 AND error_number() IN (2601, 2627) AND error_message() LIKE '%''dbo.Resource''%'
+    THROW 50409, 'Resource has been recently updated or added, please compare the resource content in code for any duplicate updates', 1;
+  ELSE
+    THROW
+END CATCH
+GO
+
+CREATE or ALTER PROCEDURE dbo.CaptureResourceIdsForChanges
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql:498]: I don't see the stored procedure UpdateResourceSearchParams. Please add.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql [hunk 4]
```diff
@@ -0,0 +1,523 @@
+/***************************************************************
+* Migration: 102
+* Description: Add new column DecompressedSize to Resource table to store the size of decompressed resource data.
+
+To avoid impacting existing functionality, the new column is added as nullable.
+To make this change both backward and forward compatible,
+A new type ResourceList_Temp will be created to include this new column.
+MergeResources SP will be changed to use both new and old types in iteration1.
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql]: 3 stored procedures should be changed, not just MergeResources.
It is worth mentioning the plan for backfilling DecompressedLength

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql [hunk 5]
```diff
@@ -14,7 +14,8 @@ CREATE TABLE dbo.CurrentResource -- This is replaced by view CurrentResource
     IsRawResourceMetaSet        bit                     NOT NULL,
     SearchParamHash             varchar(64)             NULL,
     TransactionId               bigint                  NULL,
-    HistoryTransactionId        bigint                  NULL
+    HistoryTransactionId        bigint                  NULL,
+    DecompressedSize            int                     NULL
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql]: As discussed in chat please change to DecompressedLength, as it is in line with SQL naming convention.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+   ,Version              int                 NOT NULL
+   ,HasVersionToCompare  bit                 NOT NULL -- in case of multiple versions per resource indicates that row contains (existing version + 1) value
+   ,IsDeleted            bit                 NOT NULL
+   ,IsHistory            bit                 NOT NULL
+   ,KeepHistory          bit                 NOT NULL
+   ,RawResource          varbinary(max)      NOT NULL
+   ,IsRawResourceMetaSet bit                 NOT NULL
+   ,RequestMethod        varchar(10)         NULL
+   ,SearchParamHash      varchar(64)         NULL
+   ,DecompressedSize     INT                 NULL
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql]: As discussed in chat please change to DecompressedLength, as it is in line with SQL naming convention.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs [hunk 7]
```diff
@@ -37,6 +37,7 @@ public static class SchemaVersionConstants
         public const int SearchParameterOptimisticConcurrency = (int)SchemaVersion.V95;
         public const int SearchParameterMaxLastUpdatedStoredProcedure = (int)SchemaVersion.V96;
         public const int SearchParameterLastUpdatedIndex = (int)SchemaVersion.V98;
+        public const int DecompressedSize = (int)SchemaVersion.V102;
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs]: As discussed in chat please change to DecompressedLength, as it is in line with SQL naming convention.

**Comment by SergeyGaluzo** [comment]
@rajithaalurims I think this PR is a **prerequisite** to address issue https://microsofthealth.visualstudio.com/Health/_workitems/edit/156583

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs [hunk 8]
```diff
@@ -37,6 +37,7 @@ public static class SchemaVersionConstants
         public const int SearchParameterOptimisticConcurrency = (int)SchemaVersion.V95;
         public const int SearchParameterMaxLastUpdatedStoredProcedure = (int)SchemaVersion.V96;
         public const int SearchParameterLastUpdatedIndex = (int)SchemaVersion.V98;
+        public const int DecompressedLength = (int)SchemaVersion.V102;
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs]: Probably name needs to be more specific, as there will be 3 phases.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql (hunk 1 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql]: Removed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql (hunk 2 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/MergeResources.sql]: Fixed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql (hunk 5 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql]: Changed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql (hunk 6 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql]: Changed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs (hunk 7 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs]: Changed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs (hunk 8 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs]: Changed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql (hunk 3 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/102.diff.sql:498]: Fixed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+   ,Version              int                 NOT NULL
+   ,HasVersionToCompare  bit                 NOT NULL -- in case of multiple versions per resource indicates that row contains (existing version + 1) value
+   ,IsDeleted            bit                 NOT NULL
+   ,IsHistory            bit                 NOT NULL
+   ,KeepHistory          bit                 NOT NULL
+   ,RawResource          varbinary(max)      NOT NULL
+   ,IsRawResourceMetaSet bit                 NOT NULL
+   ,RequestMethod        varchar(10)         NULL
+   ,SearchParamHash      varchar(64)         NULL
+   ,DecompressedLength     INT                 NULL
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Types/ResourceList_Temp.sql:17]: Lower case

**Review by SergeyGaluzo — DISMISSED** [review]

File: docs/arch/adr-fhir-ingested-data-size-design.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+#### Schema Changes, Resource Persistence Logic, Data Backfill
+Add DecompressedLength column to Resource table:
+-	Column: DecompressedLength INT NULL
+-	Stores the uncompressed size of each resource in bytes
+-	Nullable to support gradual rollout and historical data backfill
+  
+Parameter table entries:
+-	FHIR_TotalDataSize: Stores ( total ingested data size + total index size) in GB
+-	FHIR_TotalIndexSize: Stores total index size in GB
+-	Both entries include timestamp of last calculation
```
Comment (rbans96) [inline docs/arch/adr-fhir-ingested-data-size-design.md:24]: nit: Can we rename the entries to FHIR_TotalDataSizeInGB and FHIR_TotalIndexSizeInGB

File: docs/arch/adr-fhir-ingested-data-size-design.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+**Background Calculation Job Performance:**
+- Small database (3TB): Approximately 2 minutes per calculation cycle
+- Large database (128TB): Approximately 4 hours per calculation cycle
+- Job frequency: Runs every 4 hours to maintain current metrics
+- Database size correlation: Calculation time scales linearly with database size
+
+### Consequences
+- Background job adds periodic database load every 4 hours
+- Failure in job does not impact core FHIR server functionality
+- Falure in job results in stale data size metrics until next successful run
```
Comment (rbans96) [inline docs/arch/adr-fhir-ingested-data-size-design.md:79]: I assume we would emit failure logs from the background job. Would we need to setup any alerting/monitoring if the background job fails?
Do you think if its useful to add a section for alerts/monitors if needed at all for this design spec.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql [hunk 12]
```diff
@@ -14,7 +14,8 @@ CREATE TABLE dbo.CurrentResource -- This is replaced by view CurrentResource
     IsRawResourceMetaSet        bit                     NOT NULL,
     SearchParamHash             varchar(64)             NULL,
     TransactionId               bigint                  NULL,
-    HistoryTransactionId        bigint                  NULL
+    HistoryTransactionId        bigint                  NULL,
+    DecompressedLength            int                     NULL
```
Comment (rbans96) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql:18]: I understand this is nullable for backward compatibility. Do we have a test coverage to ensure we don't pass the null value/0 value for these changes, if it makes sense?

File: docs/arch/adr-fhir-ingested-data-size-design.md (hunk 10 above)
Comment (rajithaalurims) [reply docs/arch/adr-fhir-ingested-data-size-design.md:24]: > 

> Done

File: docs/arch/adr-fhir-ingested-data-size-design.md (hunk 11 above)
Comment (rajithaalurims) [reply docs/arch/adr-fhir-ingested-data-size-design.md:79]: Great question. I have added alerting and monitoring in the ADR I put in place on WP side. If there are no metrics being emitted as expected we do get ICMs.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql (hunk 12 above)
Comment (rajithaalurims) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/Resource.sql:18]: For iteration1 null and 0 are acceptable. After iteration 2, null values are not acceptable. Once all three iterations are done, we should not have any null values in that table. I can add test coverage in iteration 2 or 3 to make sure there are no null values.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/GetResources.sql [hunk 13]
```diff
@@ -28,6 +28,7 @@ BEGIN TRY
             ,RawResource
             ,IsRawResourceMetaSet
             ,SearchParamHash
+            ,DecompressedLength
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/GetResources.sql:31]: @rajithaalurims Why do you need to read decompressed length in the existing stored procedures?
